### DISCUSSION TITLE
Deduplication issues

### DISCUSSION THREAD

**Discussion by rhuanbarreto** [TanStack/query#3358] [General] [answered]
Regarding this issue in StackOverflow https://stackoverflow.com/questions/68482832/using-react-query-in-external-package-leads-to-no-queryclient-set-error-even-t

I also experienced the same issues and needed to run a npm dedup or yarn-deduplicate in order to solve it.

Is there anything that react-query could do to solve this dependency issue?

**Comment by TkDodo** [comment, marked as answer]
We use react context to retrieve the query client. If there are two versions of react query, each wants their own context instance, but you likely only provide one.
It depends on your setup, but if the lib that uses RQ has RQ as a peer-dependencies, it should not happen.

**Reply by rhuanbarreto** [reply]
This happens when a dependency duplication happens independently of implementation. Let's say my frontend depends on a library that depends on version `^0.10.0` and you install version `^0.16.0` on the frontend then yarn keeps both versions in node_modules and this happens. Running the deduplication command fix the issue. I don't know if such a thing could be fixed on the library level.

**Reply by TkDodo** [reply]
but that sounds weird tbh. If a library depends on `^0.10.0` and you install `^0.16.0`, it should only install `0.16.0` because that satisfies the constraint of "compatible with 0.10.0".

**Reply by rhuanbarreto** [reply]
Yes. But this happens a lot if you don't dedup packages in you lockfile. What I don't understand is even if different versions of the code are present, why don't they recognize the same provider in the code.

**Reply by TkDodo** [reply]
> But this happens a lot if you don't dedup packages in you lockfile.

I have never really done that manually. I'm mostly using `npm` but I don't think that `npm` and `yarn` would be drastically different here. It would mean _a lot of_ nested `node_modules`. I mean, everyone has a different version for `babel-runtime` in their `package.json` 😅 .

---

In one of my projects, I can se `babel-runtime` 29 times in `package-lock`, with the following versions, as `dependencies` of 3rd party libs:

- "babel-runtime": "^6.11.6", (multiple times)
- "babel-runtime": "^6.26.0",
- "babel-runtime": "6.x", (multiple times)
- "babel-runtime": "6.x.x",

What I am then getting in the top-level `node_modules` is version 6.26.0:

```
        "babel-runtime": {
            "version": "6.26.0",
            "resolved": "https://registry.npmjs.org/babel-runtime/-/babel-runtime-6.26.0.tgz",
            "integrity": "sha1-llxwWGaOgrVde/4E/yM3vItWR/4=",
```

I have checked node_modules and only have it in there one time, at the top level 🤷 

---

> why don't they recognize the same provider in the code.

because they both call `React.createContext()`, creating their own context.
